src/tools/rbd_mirror/ImageWatcher.cc

```diff
@@ -0,0 +1,151 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <boost/bind.hpp>
+#include <boost/function.hpp>
+
+#include "common/debug.h"
+#include "common/errno.h"
+#include "ImageWatcher.h"
+
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd-mirror: "
+
+using std::list;
+using std::map;
+using std::set;
+using std::string;
+using std::unique_ptr;
+using std::vector;
+
+using librados::Rados;
+using librados::IoCtx;
+using librbd::Image;
+
+namespace rbd {
+namespace mirror {
+
+namespace {
+
+struct C_RefreshImages : public Context {
+  C_RefreshImages(ImageWatcher *iw) :
+    iw(iw) {}
+  virtual void finish(int r) {
+    iw->refresh_images();
+  }
+  ImageWatcher *iw;
+};
+
+} // anonymous namespace
+
+ImageWatcher::ImageWatcher(RadosRef cluster, double interval_seconds,
+			 Mutex &lock, Cond &cond) :
+  m_lock(lock),
+  m_refresh_cond(cond),
+  m_cluster(cluster),
+  m_timer(g_ceph_context, m_lock),
+  m_interval(interval_seconds)
+{
+  m_timer.init();
+}
+
+ImageWatcher::~ImageWatcher()
+{
+  m_timer.shutdown();
+}
+
+map<int64_t, set<string> > ImageWatcher::get_images() const
+{
+  assert(m_lock.is_locked());
+  return m_images;
+}
+
+void ImageWatcher::refresh_images()
+{
+  dout(20) << __func__ << dendl;
+  map<int64_t, set<string> > images;
+  list<pair<int64_t, string> > pools;
+  int r = m_cluster->pool_list2(pools);
+  if (r < 0) {
+    derr << "error listing pools: " << cpp_strerror(r) << dendl;
+    return;
+  }
+
+  librbd::RBD rbd;
+  for (auto kv : pools) {
```

Comment:
do you think it will be too heavy? Do we need something like RBD_DIRECTORY to store mirroring or alike?